apacheconfig/parser.py

```diff
@@ -145,7 +145,7 @@ def p_config(self, p):
             p[0] = ['config', p[1]]
 
     def p_error(self, p):
-        raise ApacheConfigError("Parser error at '%s'" % p.value if p else '?')
+        raise ApacheConfigError("Parser error at '%s'" % p.value if p else 'Unexpected eof')
```

Comment:
Is this method called only on the EOF condition?

I vaguely suspect that by this point the whole source file has already been successfully tokenized to the end. Can it be that this method is called when a token can't be matched parser's BNF rules?

On top of that, that `p` can be empty e.g. not carrying any lexem. What is the condition when this can happen?

Sorry for random and vague thoughts, I just trying not to fool our users. ;-)

Reply:
I looked through the [ply-homepage](http://www.dabeaz.com/ply/ply.html#ply_nn29). Under 6.8 the first point is how p_error is called. The way I understood it is that `p` is only `None` if the syntax error occurs at the end of the file and if an error is raised the other points are neglectable. 

How would an empty `p` not carrying any lexem look like? Can there be a completely empty token? 

The problem for me was that one of the <IfModule> tags in the apache conf was never closed which just gave me an ? as error message.